### PR TITLE
Update types in classifier/

### PR DESCRIPTION
**Pull request by bhancock8**
Update type comments to reflect recent changes.

**Test plan:**
tox -e check passes and tox -e type only complains on files outside of classifier

### CODE REVIEW

File: snorkel/classification/models/advanced/advanced_classifier.py [hunk 1]
```diff
@@ -119,7 +119,7 @@ def forward(  # type: ignore
         X_dict = move_to_device(X_dict, self.config["device"])
 
         outputs: Dict[str, Mapping[Union[str, int], Any]] = {}
-        outputs["_input_"] = X_dict
+        outputs["_input_"] = X_dict  # type: ignore
```
Comment (henryre) [inline snorkel/classification/models/advanced/advanced_classifier.py]: What's the issue here?

File: snorkel/classification/models/advanced/advanced_classifier.py [hunk 2]
```diff
@@ -196,9 +199,7 @@ def calculate_loss(
         return loss_dict, count_dict
 
     @torch.no_grad()
-    def _calculate_probs(
-        self, X_dict: Mapping[Union[str, int], torch.Tensor], task_names: Iterable[str]
-    ):
+    def _calculate_probs(self, X_dict: Mapping[str, Any], task_names: Iterable[str]):
```
Comment (henryre) [inline snorkel/classification/models/advanced/advanced_classifier.py]: Not all Tensors any more?

File: snorkel/classification/scorer.py [hunk 3]
```diff
@@ -32,12 +34,14 @@ def __init__(
...
-    def score(self, golds, preds, probs):
+    def score(
+        self, golds: ArrayLike, preds: ArrayLike, probs: ArrayLike
+    ) -> Dict[str, float]:
         metric_dict = dict()
 
         for metric_name, metric in self.metrics.items():
             # Handle no examples
-            if len(golds) == 0:
+            if len(golds) == 0:  # type: ignore
```
Comment (henryre) [inline snorkel/classification/scorer.py]: Which ArrayLike has no `len`?

File: snorkel/classification/training/schedulers/scheduler.py [hunk 4]
```diff
@@ -14,7 +16,9 @@ def __init__(self):
     @abstractmethod
     def get_batches(
         self, dataloaders: Sequence[ClassifierDataLoader]
-    ) -> Tuple[Tuple[Dict[str, Any], Dict[str, Any]], ClassifierDataLoader]:
+    ) -> Iterator[
+        Tuple[Tuple[Dict[str, Any], Dict[str, Tensor]], ClassifierDataLoader]
```
Comment (henryre) [inline snorkel/classification/training/schedulers/scheduler.py]: Used below as well, let's alias this (e.g. `BatchIterator = Iterator[...]`)

File: snorkel/classification/models/advanced/advanced_classifier.py (hunk 1 above)
Comment (bhancock8) [reply snorkel/classification/models/advanced/advanced_classifier.py]: The type of outputs is: Dict[str, Mapping[Union[str, int], Any]]
The type of X_dict is: Mapping[str, Any]
 
outputs["__input__": X_dict] is unhappy that X_dict cannot possibly be Mapping[int, Any].

File: snorkel/classification/models/advanced/advanced_classifier.py (hunk 2 above)
Comment (bhancock8) [reply snorkel/classification/models/advanced/advanced_classifier.py]: Y will always be tensors. X doesn't have to be. e.g., you can pass string-based uids or other tags through as well that a module takes advantage of.

File: snorkel/classification/scorer.py (hunk 3 above)
Comment (bhancock8) [reply snorkel/classification/scorer.py]: It doesn't know what to do with scipy.sparse.

**Comment by bhancock8** [comment]
Comments addressed

File: snorkel/classification/models/advanced/advanced_classifier.py (hunk 1 above)
Comment (henryre) [reply snorkel/classification/models/advanced/advanced_classifier.py]: Try Dict[str, Union[Mapping[int, Any], Mapping[str, Any]]]

**Comment by henryre** [comment]
But see one above comment

**Comment by bhancock8** [comment]
Yeah, tried your typing suggestion @henryre, still no dice I think the issue is that we actually can't guarantee type safety here. The output of each operation in the task flow is either a dict of fields or a sequence of fields (including the trivial case of a singleton list if they just output a single tensor), and the user specifies for each operation what its inputs should be by pointing to a previously run module and either an integer index or a string key. It's possible for the user to specify an input like `(my_module_that_outputs_dicts, 0)', which would in fact fail when the task flow tries to access the 0th element of a dict. Leaving as is for now, can always revisit.
